File: 0735-asteroid-collision/0735-asteroid-collision.cs
public class Solution {
    public int[] AsteroidCollision(int[] asteroids) {
        List<int> result = new List<int>();

        foreach (int asteroid in asteroids) {
            if (result.Count == 0) {
                result.Add(asteroid);
                continue;
            }

            while (true) { 
                if (result.Count == 0) {
                    result.Add(asteroid);
                    break;
                } else {
                    int el = result[result.Count - 1];

                    
                    if (el * asteroid > 0) {
                        // same direction
                        result.Add(asteroid);
                        break;
                    } else {
                        // different direction but casefy
                        
                        if (el < 0 && asteroid > 0) {
                            // Case 1. el이 왼쪽, 행성 오른쪽 -> 충돌 가능성 없음 (break)
                            result.Add(asteroid);
                            break;
                        } else {

                            // Case 2. el이 오른쪽, 행성 왼쪽 -> 충돌 가능성 casefy

                            if (Math.Abs(el) > Math.Abs(asteroid)) {
                                // el이 더 큰 경우
                                break;
                            } else if (Math.Abs(el) < Math.Abs(asteroid)) {
                                // 행성이 더 큰 경우
                                result.RemoveAt(result.Count - 1);
                                continue;
                            } else {
                                // 똑같은 경우
                                result.RemoveAt(result.Count - 1);
                                break;   
                            }
                        }

                    }

                }
            }
        }
        
        return result.ToArray();
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.swift
class Solution {
    func asteroidCollision(_ asteroids: [Int]) -> [Int] {
        
        var result = [Int]()
    
        for asteroid in asteroids {
            if result.isEmpty { 
                result.append(asteroid)     
                continue
            }
            
            while true {
                guard let el = result.popLast() else {
                    result.append(asteroid)
                    break
                }
                // el vs asteroid
                
                // 방향이 같은 경우 
                // el 먼저 넣고  asteroid 넣고 break 하기
                
                if el * asteroid > 0 { 
                    result.append(el)
                    result.append(asteroid)
                    break
                }
                
                // 방향이 다른 경우는 case를 따져야 함
                else {
                    
                
                // el이 왼쪽(-) 으로 가고, asteroid (+)으로 가는 경우 -> 충돌 안함
                // el 먼저 넣고, asteroid 넣고 break
                    
                    if el < 0 && asteroid > 0 { 
                        result.append(el)
                        result.append(asteroid)
                        break
                    }
                    
                    // el이 오른쪽(+) 으로 가고, asteroid (-)으로 가는 경우 -> 충돌함
                    // 기존의 것(el)이 더 쎄면 el 넣고 break
                    // 새로운 것(asteroid)이 더 쎄면 asteroid 넣고 continue
                    // 같은 경우 둘다 파괴하고 break

                    else {
                        
                        if abs(el) > abs(asteroid) { 
                            result.append(el)
                            break
                        } else if abs(el) < abs(asteroid) {
                            continue
                        }
                        else {
                            break
                        }
                    }
                    
                }
                
            }
            
            
        }
    
        
        return result
        
    }
}